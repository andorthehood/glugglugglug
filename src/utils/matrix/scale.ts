import { multiply } from "./multiply";
import { scaling } from "./scaling";

/**
 * Multiplies a matrix by a scaling transform.
 * @param m Matrix to transform.
 * @param sx Scale factor along the x axis.
 * @param sy Scale factor along the y axis.
 * @param sz Scale factor along the z axis.
 * @param dst Destination matrix to write into. Defaults to mutating the supplied matrix.
 * @returns The transformed matrix.
 */
const scalingMatrix = new Array<number>(16);

export function scale(
	m: number[],
	sx: number,
	sy: number,
	sz: number,
	dst: number[] = m,
) {
	scaling(sx, sy, sz, scalingMatrix);
	return multiply(m, scalingMatrix, dst);
}
