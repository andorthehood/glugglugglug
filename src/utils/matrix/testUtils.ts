export function roundValue(value: number, precision = 6) {
	const factor = 10 ** precision;
	const rounded = Math.round(value * factor) / factor;
	return Object.is(rounded, -0) ? 0 : rounded;
}

export function formatMatrix(values: number[]) {
	return Array.from({ length: 4 }, (_, row) => {
		const start = row * 4;
		return values.slice(start, start + 4).map((value) => roundValue(value));
	});
}

export function formatVector(values: number[]) {
	return values.map((value) => roundValue(value));
}
