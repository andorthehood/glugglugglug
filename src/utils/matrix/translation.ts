/**
 * Creates a 4x4 matrix that translates by the supplied offsets.
 * @param tx Translation along the x axis.
 * @param ty Translation along the y axis.
 * @param tz Translation along the z axis.
 * @param dst Destination matrix to write into. When omitted a new matrix is created.
 * @returns A translation matrix.
 */
export function translation(
	tx: number,
	ty: number,
	tz: number,
	dst: number[] = new Array(16),
) {
	dst[0] = 1;
	dst[1] = 0;
	dst[2] = 0;
	dst[3] = 0;
	dst[4] = 0;
	dst[5] = 1;
	dst[6] = 0;
	dst[7] = 0;
	dst[8] = 0;
	dst[9] = 0;
	dst[10] = 1;
	dst[11] = 0;
	dst[12] = tx;
	dst[13] = ty;
	dst[14] = tz;
	dst[15] = 1;
	return dst;
}
