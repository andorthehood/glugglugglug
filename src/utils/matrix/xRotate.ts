import { multiply } from "./multiply";
import { xRotation } from "./xRotation";

/**
 * Multiplies a matrix by a rotation around the x axis.
 * @param m Matrix to transform.
 * @param angleInRadians Rotation angle in radians.
 * @param dst Destination matrix to write into. Defaults to mutating the supplied matrix.
 * @returns The transformed matrix.
 */
const rotationMatrix = new Array<number>(16);

export function xRotate(
	m: number[],
	angleInRadians: number,
	dst: number[] = m,
) {
	xRotation(angleInRadians, rotationMatrix);
	return multiply(m, rotationMatrix, dst);
}
