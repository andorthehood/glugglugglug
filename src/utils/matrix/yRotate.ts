import { multiply } from "./multiply";
import { yRotation } from "./yRotation";

/**
 * Multiplies a matrix by a rotation around the y axis.
 * @param m Matrix to transform.
 * @param angleInRadians Rotation angle in radians.
 * @param dst Destination matrix to write into. Defaults to mutating the supplied matrix.
 * @returns The transformed matrix.
 */
const rotationMatrix = new Array<number>(16);

export function yRotate(
	m: number[],
	angleInRadians: number,
	dst: number[] = m,
) {
	yRotation(angleInRadians, rotationMatrix);
	return multiply(m, rotationMatrix, dst);
}
