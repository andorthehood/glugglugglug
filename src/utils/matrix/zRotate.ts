import { multiply } from "./multiply";
import { zRotation } from "./zRotation";

/**
 * Multiplies a matrix by a rotation around the z axis.
 * @param m Matrix to transform.
 * @param angleInRadians Rotation angle in radians.
 * @param dst Destination matrix to write into. Defaults to mutating the supplied matrix.
 * @returns The transformed matrix.
 */
const rotationMatrix = new Array<number>(16);

export function zRotate(
	m: number[],
	angleInRadians: number,
	dst: number[] = m,
) {
	zRotation(angleInRadians, rotationMatrix);
	return multiply(m, rotationMatrix, dst);
}
