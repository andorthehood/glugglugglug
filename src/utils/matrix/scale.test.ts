import { describe, expect, it } from "vitest";
import { scale } from "./scale";
import { translation } from "./translation";
import { xRotate } from "./xRotate";
import { yRotate } from "./yRotate";
import { zRotate } from "./zRotate";
import { formatMatrix } from "./testUtils";

describe("scale", () => {
	it("composes transforms via helpers", () => {
		const base = translation(5, -3, 2);
		const composed = scale(
			zRotate(
				yRotate(xRotate(base, Math.PI / 6), Math.PI / 4),
				Math.PI / 3,
			),
			2,
			2,
			2,
		);

		expect(composed).toBe(base);
		expect(formatMatrix(composed)).toMatchSnapshot();
	});

	it("supports writing into a separate destination", () => {
		const base = translation(1, 2, 3);
		const dst = new Array<number>(16);
		const result = scale(base, 2, 3, 4, dst);

		expect(result).toBe(dst);
		expect(formatMatrix(result)).toMatchSnapshot();
		expect(formatMatrix(base)).toEqual(formatMatrix(translation(1, 2, 3)));
	});
});
