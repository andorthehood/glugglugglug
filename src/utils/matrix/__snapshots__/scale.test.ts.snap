// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`scale > composes transforms via helpers 1`] = `
[
  [
    0.707107,
    1.853553,
    0.253653,
    0,
  ],
  [
    -1.224745,
    0.253653,
    1.56066,
    0,
  ],
  [
    1.414214,
    -0.707107,
    1.224745,
    0,
  ],
  [
    5,
    -3,
    2,
    1,
  ],
]
`;

exports[`scale > supports writing into a separate destination 1`] = `
[
  [
    2,
    0,
    0,
    0,
  ],
  [
    0,
    3,
    0,
    0,
  ],
  [
    0,
    0,
    4,
    0,
  ],
  [
    1,
    2,
    3,
    1,
  ],
]
`;
